// @coder/node-browser/out/common/connection.js
/* eslint-disable @typescript-eslint/no-explicit-any */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var DefaultLogger = /** @class */ (function () {
        function DefaultLogger(name) {
            this.name = name;
        }
        DefaultLogger.prototype.trace = function (message) {
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            if (typeof process !== "undefined" && process.env && process.env.LOG_LEVEL === "trace") {
                this.log.apply(this, [message].concat(args));
            }
        };
        DefaultLogger.prototype.error = function (message) {
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            this.log.apply(this, [message].concat(args));
        };
        DefaultLogger.prototype.log = function (message) {
            var args = [];
            for (var _i = 1; _i < arguments.length; _i++) {
                args[_i - 1] = arguments[_i];
            }
            console.log.apply(console, ["[" + this.name + "]", message].concat(args.map(function (a) { return JSON.stringify(typeof a === "function" ? a() : a); })));
        };
        return DefaultLogger;
    }());
    exports.DefaultLogger = DefaultLogger;
    var ConnectionStatus;
    (function (ConnectionStatus) {
        ConnectionStatus[ConnectionStatus["Disconnected"] = 0] = "Disconnected";
        ConnectionStatus[ConnectionStatus["Connected"] = 1] = "Connected";
        ConnectionStatus[ConnectionStatus["Closed"] = 2] = "Closed";
    })(ConnectionStatus = exports.ConnectionStatus || (exports.ConnectionStatus = {}));
});
